_posts/2019-12-25-technical-debt.md

```diff
@@ -0,0 +1,51 @@
+
+It is impossible to measure the intangible technical debt, but this property does not always extend to its effects. We can find another measurement (or more) that can act as a suitable proxy. This is not too dissimilar to maintenance practices in heavy industry. You can never know for certain when a piece of equipment is going to fail, but you can measure how well maintenance crews are adhering to a planned maintenance schedule. In most cases this is a reasonable indication of equipment health*. What measures could we identify that could act as reasonable proxies for the amount of technical debt across one or more software systems?
+
+I believe a suitable proxy for technical debt can be found in existing literature on software delivery. The State of DevOps Report produced by DevOps Research and Assessment (hereby refered to as "the DORA Report") is compiled annually and details the habits of high performing software teams. It is one of the most rigourous studies on software development, produced from surveying and observing hundreds of teams across the world. If we are to come upon a satisfactory measure of technical debt, it would likely be in the DORA report. The report distinguishes two key measures of software delivery performance; throughput and stability. Throughput is measured via deployment frequency and lead time for changes. Stability is measured via time to restore a service and the change failure rate. I believe that change lead time and time to restore service are useful signals of the level of technical debt a project has acrued. Both of these marry well with the effects of technical debt discussed earlier, which included decreased velocity and increasingly risky changes.* Both of these measures are also difficult to game.*
+
+TODO: Why is change lead time a good measure
+
+TODO: Why is MTTR a good measure
+
+This raises an interesting question to me; if software requires no additional features, and is never in need of repair, is it free of technical debt? Imagine two bodies of work. One has the former traits, along with the reputation for having been terrible to work with by the authors. The other is a more mediocre example of software development, currently in active development and incurring under 5 hours of downtime per year. Which suggests the lesser liability? If I fixate on the numbers in the shadow of our prior definition, I would be led to say the first system has less technical debt. But I guarantee that many would argue in favour of the second. There is an element of potential within technical debt; the absolute zero of which only occurs when the last change to be made is to turn it off.
```

Comment:
Very existential. If a tree falls in the forest... ?

Reply:
The idea was to explore a proof via contradiction.

Reply:
I like your idea, but currently reads poorly - simplify the language to make your point clearly.

You state your preference - but don't supply justification/line of reasoning - turn it into a concrete example.